src/cabal-details.js

```diff
@@ -0,0 +1,365 @@
+  }
+
+  getLocalUser() {
+    return this.user
+  }
+
+  getLocalName() {
+      return this.user.name || this.user.key.slice(0,8)
+  }
+  
+  joinChannel(channel) {
+    var details = this.channels[channel]
+    // we created a channel
+    if (!details) {
+        details = new ChannelDetails(this._cabal, channel)
+        this.channels[channel] = details
+    }
+    // we weren't already in the channel, join
+    if (!details.join()) { 
+      var joinMsg = {
+        type: "channel/join",
+        content: { channel }
+      }
+      // publish a join message to the cabal to signify our presence
+      this._cabal.publish(joinMsg)
+    }
+    // we probably always want to open a joined channel?
+    this.openChannel(channel)
+  }
+
+  leaveChannel(channel) {
+    if (!channel) channel = this.chname
+    if (channel === "!status") return
+    var joined = this.getJoinedChannels()
+    var details = this.channels[channel]
+    if (!details) return
+    var left = details.leave()
+    // we were in the channel, leave
+    if (left) { 
+      var leaveMsg = {
+        type: "channel/leave",
+        content: { channel }
+      }
+      this._cabal.publish(leaveMsg)
+    }
+    var indexOldChannel = joined.indexOf(channel)
+    var newChannel
+    // open up another channel if we left the one we were viewing
+    if (channel === this.chname) {
+      let newIndex = indexOldChannel + 1 
+      if (indexOldChannel >= joined.length) newIndex = 0
+      newChannel = joined[newIndex] || "!status"
+    }
+    this.closeChannel(channel, newChannel)
+  }
+
+  getUsers() {
+    return this.users
+  }
+
+  _emitUpdate() {
+    this.emit('update', this)
+  }
+
+  registerListener(source, event, listener) {
+    this.listeners.push({ source, event, listener })
+    source.on(event, listener)
+  }
+
+  _destroy () {
+    this.listeners.forEach((obj) => { obj.source.removeListener(obj.event, obj.listener)})
+  }
+
+  _initializeUser() {
+    this._cabal.getLocalKey((err, lkey) => {
+      if (err) throw err
+      this.user.key = lkey
+      this.user.local = true
+      this.user.online = true
+      this.users[lkey] = this.user
+      // try to get more data for user
+      this._cabal.users.get(lkey, (err, user) => {
+        if (!user) { return }
+        this.user = user
+        // restore `user.local` and `user.online` as they don't come from cabal-core
+        this.user.local = true
+        this.user.online = true
+        this._emitUpdate()
+      })
+    })
+  }
+
+  _initialize() {
+    const cabal = this._cabal
+    // populate channels
+    cabal.channels.get((err, channels) => {
+      channels.forEach((channel) => {
+        let details = this.channels[channel]
+        if (!details) {
+          this.channels[channel] = new ChannelDetails(cabal, channel)
+        }
+        // listen for updates that happen within the channel
+        cabal.messages.events.on(channel, this.messageListener.bind(this))
```

Comment:
Do we un-listen when this cabal is destroyed?

Reply:
hm, how would we detect that a cabal is destroyed?

we have `_destroy()` [CabalDetails line 215](https://github.com/cabal-club/cabal-client/pull/11/files/3ebde46532ce1722b0fa34e24bcb66e1946b6570#diff-29fea628f7f8cdba0f19b72b4fb6ba86R215) which is used when removeCabal is called from Client.